File: BST.py
# import pickle


class Node:
    def __init__(self, n):
        self.left = None
        self.right = None
        self.height = 1
        self.slope = 0
        self.parent = None
        try:
            if len(n) > 0:
                self.data = n[0]
                for r in range(1, len(n)):
                    self.insert(n[r])
        except:
            self.data = n

    def rightrotation(self):
        a = self.data
        b = self.right.data
        node1 = self.left
        node2 = self.right.left
        node3 = self.right.right
        self.data = b
        self.left = Node(a)
        self.left.parent = self
        self.right = node3
        if node3 is not None:
            node3.parent = self
            node3.heightfunc()
            node3.slopefunc(setvalue=True)
        self.left.left = node1
        if node1 is not None:
            node1.parent = self.left
            node1.heightfunc()
            node1.slopefunc(setvalue=True)
        self.left.right = node2
        if node2 is not None:
            node2.parent = self.right
            node2.heightfunc()
            node2.slopefunc(setvalue=True)
        self.left.heightfunc()
        self.heightfunc()
        self.slopefunc(setvalue=True)
        self.left.slopefunc(setvalue=True)
        #print(self.height)

    def balance(self):
        if self.slope < -1:
            self.rightrotation()

    def heightfunc(self, nodeparam=None):
        temp = self if nodeparam is None else nodeparam
        leftheight = 0
        rightheight = 0
        while temp.left is not None:
            temp = temp.left
            leftheight += 1
        temp = self if nodeparam is None else nodeparam
        while temp.right is not None:
            temp = temp.right
            rightheight += 1
        if nodeparam is not None:
            return max(leftheight + 1, rightheight + 1)
        self.height = max(leftheight + 1, rightheight + 1)

        # print(self.height)

    def slopefunc(self, nodeparam=None, returnvalue=None, setvalue=None):
        returnvalue = returnvalue if returnvalue is not None else False
        setvalue = setvalue if setvalue is not None else False
        temp = self if nodeparam is None else nodeparam
        if temp.right is not None and temp.left is not None:
            # print(temp.data, temp.left.height, temp.right.height)
            slopep = temp.left.height - temp.right.height
        elif temp.right is not None:
            slopep = -temp.right.height
        elif temp.left is not None:
            slopep = temp.left.height
        else:
            slopep = 0
        if setvalue:
            self.slope = slopep
        if returnvalue:
            return slopep

    def find(self, num, returnobj=None):
        returnobj = returnobj if returnobj is not None else False
        temp = self
        while temp is not None:
            if num < temp.data:
                temp = temp.left
            elif num > temp.data:
                temp = temp.right
            elif temp.data == num:
                if returnobj:
                    return temp
                else:
                    return 1
        else:
            if returnobj:
                return None
            else:
                return 0

    def insert(self, n):
        if n < self.data:
            if self.left is None:
                self.left = Node(n)
                self.left.parent = self
                self.heightfunc()
                self.slope = self.slopefunc(returnvalue=True)
            else:
                self.left.insert(n)
                self.heightfunc()
                self.slope = self.slopefunc(returnvalue=True)
                self.balance()
                self.heightfunc()
        elif n > self.data:
            if self.right is None:
                self.right = Node(n)
                self.right.parent = self
                self.heightfunc()
                self.slope = self.slopefunc(returnvalue=True)
            else:
                self.right.insert(n)
                self.heightfunc()
                self.slope = self.slopefunc(returnvalue=True)
                self.balance()
                self.heightfunc()

    def inOrder(self):
        print(self.data, self.height, self.slope)
        if self.left is not None:
            self.left.inOrder()
        if self.right is not None:
            self.right.inOrder()


root = Node([6,3,2,1,8])
root.inOrder()

'''for i in range(19,0,-1):
    root.insert(i)
    root.inOrder()
    print()'''

File: BST.cpp
#include<iostream>
using namespace std;
class Node
{
	int data,height;
	Node *left,*right,*parent;
	void heightCalc(Node *temp)
	{
		if(temp->left != NULL && temp->right != NULL)
			temp->height = (temp->left->height>temp->right->height)?(1+temp->left->height):(1+temp->right->height);
		else if(temp->left != NULL)
			temp->height = 1+temp->left->height;
		else if(temp->right != NULL)
			temp->height = 1+temp->right->height;
		else
			temp->height = 1;
	}
	int slope(Node *temp)
	{
		if(temp->left != NULL && temp->right != NULL)
			return temp->left->height-temp->right->height;
		else if(temp->left != NULL)
			return temp->left->height;
		else if(temp->right != NULL)
			return 0-temp->right->height;
		return 0;

	}
	void rightRotate(Node *temp)
	{
		int a,b;
		Node *n1,*n2,*n3;
		a = temp->data;
		b = temp->left->data;
		n1 = temp->left->left;
		n2 = temp->left->right;
		n3 = temp->right;
		
		temp->data = b;
		temp->right = temp->left;
		temp->right->data = a;
		temp->left = n1;
		if(n1 != NULL)
			temp->left->parent = temp;
		temp->right->left = n2;
		if(n2 != NULL)
		    temp->right->left->parent = temp->right;
		temp->right->right = n3;
		if(n3 != NULL)
			temp->right->right->parent = temp->right;
		heightCalc(temp->right);
		heightCalc(temp);
		//cout<<temp->right->height<<endl;
	}
	void leftRotate(Node *temp)
	{
		int a,b;
		Node *n1,*n2,*n3;
		a = temp->data;
		b = temp->right->data;
		n1 = temp->left;
		n2 = temp->right->left;
		n3 = temp->right->right;
		
		temp->data = b;
		//temp->right = new Node(a);
		temp->left = temp->right;
		temp->left->data = a;
		//temp->right->parent = temp;
		temp->right = n3;
		if(n3 != NULL)
			temp->right->parent = temp;
		temp->left->left = n1;
		if(n1 != NULL)
			temp->left->left->parent = temp->left;
		temp->left->right = n2;
		if(n2 != NULL)
			temp->left->right->parent = temp->left;	
		heightCalc(temp->left);
		heightCalc(temp);
	}
	void balance(Node *temp)
	{
		if(slope(temp)==+2)
		{
			if(slope(temp->left)==-1)
				leftRotate(temp->left);
			rightRotate(temp);
		}
		else if(slope(temp)==-2)
		{
			if(slope(temp->right)==1)
				rightRotate(temp->right);
			leftRotate(temp);
		}
	}
	void deleten(Node *temp)
	{
		if(temp->left == NULL && temp->right == NULL)
    		deleteLeaf(temp);
		else if(temp->left != NULL && temp->right == NULL)
			deleteOneChild(temp,0);
		else if(temp->left == NULL && temp->right != NULL)
			deleteOneChild(temp,1);
		else if(temp->left != NULL && temp->right != NULL)
			deleteChild(temp);
	}
	void heightBalDel(Node *temp)
	{
		while(temp->parent != NULL)
		{
			balance(temp);
			heightCalc(temp);
			temp = temp->parent;
		}
		balance(temp);
		heightCalc(temp);
	}
	void deleteLeaf(Node *temp)
	{
		if(temp->data > temp->parent->data)
		{
			temp->parent->right = NULL;
			heightBalDel(temp->parent);
			temp->parent = NULL;
		}
		else if(temp->data < temp->parent->data)
		{
			temp->parent->left = NULL;
			heightBalDel(temp->parent);
			temp->parent = NULL;
		}
	}
	void deleteOneChild(Node *temp,int ch)
	{
		if(ch == 0)
		{
			if(temp->data > temp->parent->data)
			{
				temp->parent->right = temp->left;
				heightBalDel(temp->parent);
				temp->parent = NULL;
				temp->left = NULL;
			}
			else if(temp->data < temp->parent->data)
			{
				temp->parent->left = temp->left;
				heightBalDel(temp->parent);
				temp->parent = NULL;
				temp->left = NULL;
			}
		}
		else if(ch == 1)
		{
			if(temp->data > temp->parent->data)
			{
				temp->parent->right = temp->right;
				heightBalDel(temp->parent);
				temp->parent = NULL;
				temp->right = NULL;
			}
			else if(temp->data < temp->parent->data)
			{
				temp->parent->left = temp->right;
				heightBalDel(temp->parent);
				temp->parent = NULL;
				temp->right = NULL;
			}
		}
	}
    void deleteChild(Node *temp)
    {
    	Node *pr;
    	pr = predn(temp->data);
    	temp->data = pr->data;
    	deleten(pr);
	}
	Node *findn(int t)
	{
		Node *temp = this;
		while(temp != NULL)
		{
			if(temp->data == t)
			    return temp;
			else if(t<temp->data)
			    temp = temp->left;
			else if(t>temp->data)
			    temp = temp->right;
		}
		return NULL;
    }
	Node *predn(int n)
	{
		Node *temp; 
		temp = findn(n);
		if(temp->left != NULL)
			return maxin(temp->left);
		else
		{
			while(temp->parent != NULL && temp == temp->parent->left)
			{
				temp = temp->parent;
			}
		}
		return temp->parent;
	}
	Node *succn(int n)
	{
		Node *temp;
		temp = findn(n);
		if(temp->right != NULL)
			return minin(temp->right);
		else
		{
			while(temp->parent != NULL && temp == temp->parent->right)
			{
				temp = temp->parent;
			}
		}
		return temp->parent;
	}
	Node *maxin(Node *temp)
	{
		while(temp->right != NULL)
		    temp = temp->right;
		return temp;
	}
	Node *minin(Node *temp)
	{
		while(temp->left != NULL)
		    temp = temp->left;
		return temp;
	}
	public:
		Node(int d)
		{
			this->data = d;
			height = 1;
			left = NULL;
			right = NULL;
			parent = NULL;
		}
		int maxi(Node *temp)
		{
			while(temp->right != NULL)
			    temp = temp->right;
			return temp->data;
		}
		int mini(Node *temp)
		{
			while(temp->left != NULL)
			    temp = temp->left;
			return temp->data;
		}
		int pred(int n)
		{
			Node *temp; 
			temp = findn(n);
			if(temp->left != NULL)
				return maxi(temp->left);
			else
			{
				while(temp->parent != NULL && temp == temp->parent->left)
				{
					temp = temp->parent;
				}
			}
			return temp->parent->data;
		}
		int succ(int n)
		{
			Node *temp;
			temp = findn(n);
			if(temp->right != NULL)
				return mini(temp->right);
			else
			{
				while(temp->parent != NULL && temp == temp->parent->right)
				{
					temp = temp->parent;
				}
			}
			return temp->parent->data;
		}
		void insert(int d)
		{
			if(d < this->data)
			{
			    if(left == NULL)
			    {
			    	left = new Node(d);
			    	left->parent = this;
					heightCalc(this);
			    	//cout<<"lheight:"<<height<<data<<endl;
				}
				else
				{
					left->insert(d);
					balance(this);
					heightCalc(this);
					//cout<<"lheight:"<<height<<data<<endl;
				}
			}
			else if(d > this->data)
			{
				if(right == NULL)
				{
					right = new Node(d);
					right->parent = this;
					heightCalc(this);
					//cout<<"rheight:"<<height<<data<<endl;
				}
				else
				{
					right->insert(d);
					balance(this);
					heightCalc(this);
					//cout<<"rheight:"<<height<<data<<endl;
				}
			}
		}
		int find(int t)
		{
			Node *temp = this;
			while(temp != NULL)
			{
				if(temp->data == t)
				    return 1;
				else if(t<temp->data)
				    temp = temp->left;
				else if(t>temp->data)
				    temp = temp->right;
			}
			return 0;
	    }
	    void deleteN(int n)
	    {
	    	Node *temp;
	    	temp = findn(n);
	    	deleten(temp);
		}
		void inOrder()
		{
			if(left != NULL)
			{
				left->inOrder();	
			}
			cout<<this->data<<" "<<this->height<<endl;
			if(right != NULL)
			{
				right->inOrder();
			}	
		}
		
};
main()
{
	Node root(7);
	root.insert(4);
	root.insert(9);
	root.insert(1);
	root.insert(2);
	root.insert(8);
	root.insert(10);
	root.deleteN(4);
	root.inOrder();
    
}
